language/diem-framework/modules/Vote.move

```diff
@@ -0,0 +1,346 @@
+/// The Vote module is used to allow voting on proposals on the chain.
+/// It is typically not going to be used directly, but is intended to be
+/// used as a library for modules which want to perform operations which
+/// require m-of-n approvals from accounts on chain.
+/// A typical workflow would look like the following
+/// * Module M creates a ballot with a given `Proposal` and an approval policy using `create_ballot`
+/// * It receives the BallotID corresponding to the ballot
+/// * It submits votes using the `vote` function from the voters
+/// * If a vote causes a ballot to be approved, `vote` returns `true` and Module M can proceed with the operation requested by the `Proposal`
+module DiemFramework::Vote {
+
+    use Std::BCS;
+    use Std::Errors;
+    use Std::Event;
+    use Std::Signer;
+    use Std::Vector;
+    use DiemFramework::DiemTimestamp;
+
+    /// An unique identifier for a ballot. A counter is stored
+    /// under each proposers address which is incremented
+    /// every time a new ballot is created. The proposers
+    /// address is also part of the ballot id.
+    struct BallotID has store, copy, drop {
+        counter: u64,
+        proposer: address,
+    }
+
+    /// Ballot is a struct which contains a Proposal on which
+    /// votes are gathered. A ballot is started by a proposer
+    /// and it carries the proposal, the approval policy,
+    /// start/expiration timestamps
+    struct Ballot<Proposal: store + drop> has store, copy, drop {
+        /// Details for the proposal being voted on.
+        proposal: Proposal,
+        /// A human readable type for the proposal - ex: "create_validator_owner",
+        /// "freeze_account", "create_vasp", etc.
+        /// This lives outside the `proposalProposal` to make it easy for
+        /// indexers to index the ballots which have been proposed and
+        /// categorize them into "types"
+        proposal_type: vector<u8>,
+        /// The num_votes_required for this proposal to be approved
+        num_votes_required: u64,
+        /// A vector of addresses which are allowed to vote on this ballot.
+        /// Duplicates are allowed. Duplicates represent the "weight" of a voter
```

Comment:
Yeah, we don't have tuples in move. Alternatively, it can be done via a simple named struct.

```
WeightedVoter {
  weight: u64,
  voter: vector<u8>,
}
```

Reply:
Oh, if we're allowed to have something like `WeightedVoter`, then I think it'd definitely simplify things! :)

Reply:
Updated to use `WeightedVoter`